src/filters/cosmetic.rs

```diff
@@ -0,0 +1,1493 @@
+                    Ok(x) => hostname.push_str(&x),
+                    Err(_) => return Err(CosmeticFilterError::PunycodeError),
+                }
+            }
+            let negation = hostname.starts_with('~');
+            let entity = hostname.ends_with(".*");
+            let start = if negation {
+                1
+            } else {
+                0
+            };
+            let end = if entity {
+                hostname.len() - 2
+            } else {
+                hostname.len()
+            };
+            let hash = crate::utils::fast_hash(&hostname[start..end]);
+            match (negation, entity) {
+                (true, true) => not_entities_vec.push(hash),
+                (true, false) => not_hostnames_vec.push(hash),
+                (false, true) => entities_vec.push(hash),
+                (false, false) => hostnames_vec.push(hash),
+            }
+        };
+
+        /// Sorts `vec` and wraps it in `Some` if it's not empty, or returns `None` if it is.
+        #[inline]
+        fn sorted_or_none<T: std::cmp::Ord>(mut vec: Vec<T>) -> Option<Vec<T>> {
+            if!vec.is_empty() {
+                vec.sort();
+                Some(vec)
+            } else {
+                None
+            }
+        }
+
+        let entities = sorted_or_none(entities_vec);
+        let hostnames = sorted_or_none(hostnames_vec);
+        let not_entities = sorted_or_none(not_entities_vec);
+        let not_hostnames = sorted_or_none(not_hostnames_vec);
+
+        Ok((entities, not_entities, hostnames, not_hostnames))
+    }
+
+    /// Parses the contents of a cosmetic filter rule following the `##` or `#@#` separator.
+    ///
+    /// On success, updates the contents of `selector` and `style` according to the rule.
+    ///
+    /// This should only be called if the rule part after the separator has been confirmed not to
+    /// be a script injection rule using `+js()`.
+    #[inline]
+    fn parse_after_sharp_nonscript<'a>(
+        line: &'a str,
+        suffix_start_index: usize,
+        selector: &mut &'a str,
+        style: &mut Option<String>
+    ) -> Result<(), CosmeticFilterError> {
+        let mut index_after_colon = suffix_start_index;
+        while let Some(colon_index) = line[index_after_colon..].find(':') {
+            let colon_index = colon_index + index_after_colon;
+            index_after_colon = colon_index + 1;
+            if line[index_after_colon..].starts_with("style") {
+                if line.chars().nth(index_after_colon + 5) == Some('(') && line.chars().nth(line.len() - 1) == Some(')') {
+                    *selector = &line[suffix_start_index..colon_index];
+                    *style = Some(line[index_after_colon + 6..line.len()-1].to_string());
+                } else {
+                    return Err(CosmeticFilterError::InvalidStyleSpecifier);
+                }
+            } else if line[index_after_colon..].starts_with("-abp-")
```

Comment:
not on the critical path and not checked what compiler maps it to, but could store `line[index_after_colon..]` in a variable and use that instead of repeating